File: app/src/main/java/com/trevorwiebe/trackacow/presentation/activities/ArchivesActivity.java
package com.trevorwiebe.trackacow.presentation.activities;

import android.content.Intent;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.domain.adapters.ArchiveRvAdapter;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.archivedLot.QueryArchivedLots;
import com.trevorwiebe.trackacow.data.entities.ArchivedLotEntity;
import com.trevorwiebe.trackacow.domain.utils.Constants;
import com.trevorwiebe.trackacow.domain.utils.ItemClickListener;
import com.trevorwiebe.trackacow.presentation.lot_reports.LotReportActivity;

import java.util.ArrayList;

public class ArchivesActivity extends AppCompatActivity implements
        QueryArchivedLots.OnArchivedLotsLoaded {

    private ArchiveRvAdapter archiveRvAdapter = new ArchiveRvAdapter();
    private ArrayList<ArchivedLotEntity> archivedLotEntities = new ArrayList<>();

    private TextView mNoArchives;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_archives);

        mNoArchives = findViewById(R.id.no_archives);
        RecyclerView archiveRv = findViewById(R.id.archive_rv);
        archiveRv.setLayoutManager(new LinearLayoutManager(this));
        archiveRv.setAdapter(archiveRvAdapter);

        archiveRv.addOnItemTouchListener(new ItemClickListener(this, archiveRv, new ItemClickListener.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                ArchivedLotEntity archivedLotEntity = archivedLotEntities.get(position);
                Intent archiveReportsIntent = new Intent(ArchivesActivity.this, LotReportActivity.class);
                archiveReportsIntent.putExtra("lotId", archivedLotEntity.getLotId());
                archiveReportsIntent.putExtra("reportType", Constants.ARCHIVE);
                startActivity(archiveReportsIntent);
            }

            @Override
            public void onLongItemClick(View view, int position) {

            }
        }));

        new QueryArchivedLots(ArchivesActivity.this).execute(ArchivesActivity.this);

    }

    @Override
    public void onArchivedLotsLoaded(ArrayList<ArchivedLotEntity> archivedLotEntities) {
        this.archivedLotEntities = archivedLotEntities;
        if (archivedLotEntities.size() == 0) {
            mNoArchives.setVisibility(View.VISIBLE);
        } else {
            mNoArchives.setVisibility(View.INVISIBLE);
        }
        archiveRvAdapter.swapArchivedLots(archivedLotEntities);
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/archives/ArchivesActivity.kt
package com.trevorwiebe.trackacow.presentation.archives

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.trevorwiebe.trackacow.R
import com.trevorwiebe.trackacow.domain.models.lot.LotModel
import com.trevorwiebe.trackacow.domain.utils.Constants
import com.trevorwiebe.trackacow.domain.utils.DataSource
import com.trevorwiebe.trackacow.domain.utils.ItemClickListener
import com.trevorwiebe.trackacow.presentation.lot_reports.LotReportActivity
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class ArchivesActivity : AppCompatActivity() {

    private val archiveRvAdapter = ArchiveRvAdapter()
    private var archivedLotModels: List<LotModel> = emptyList()
    private lateinit var mNoArchives: TextView
    private lateinit var mProgressIndicator: LinearProgressIndicator

    private val mArchivesViewModel: ArchivesViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_archives)

        mNoArchives = findViewById(R.id.no_archives)
        mProgressIndicator = findViewById(R.id.archives_progress_indicator)
        val archiveRv = findViewById<RecyclerView>(R.id.archive_rv)
        archiveRv.layoutManager = LinearLayoutManager(this)
        archiveRv.adapter = archiveRvAdapter
        archiveRv.addOnItemTouchListener(
            ItemClickListener(
                this,
                archiveRv,
                object : ItemClickListener.OnItemClickListener {
                    override fun onItemClick(view: View, position: Int) {
                        val lotModel = archivedLotModels[position]
                        val archiveReportsIntent =
                            Intent(this@ArchivesActivity, LotReportActivity::class.java)
                        archiveReportsIntent.putExtra("lotId", lotModel.lotPrimaryKey)
                        archiveReportsIntent.putExtra("reportType", Constants.ARCHIVE)
                        archiveReportsIntent.putExtra(
                            "lotCloudDatabaseId",
                            lotModel.lotCloudDatabaseId
                        )
                        startActivity(archiveReportsIntent)
                    }

                    override fun onLongItemClick(view: View, position: Int) {}
                })
        )

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.RESUMED) {
                mArchivesViewModel.uiState.collect {

                    if (it.isFetchingFromCloud && it.dataSource == DataSource.Local) {
                        mProgressIndicator.visibility = View.VISIBLE
                    } else {
                        mProgressIndicator.visibility = View.GONE
                    }

                    archivedLotModels = it.archivedLotList

                    if (archivedLotModels.isEmpty()) {
                        mNoArchives.visibility = View.VISIBLE
                    } else {
                        mNoArchives.visibility = View.INVISIBLE
                    }
                    archiveRvAdapter.swapArchivedLots(archivedLotModels)
                }
            }
        }
    }
}